### PR TITLE
Add ConsoloniaAccessText underline

### PR DESCRIPTION
**Pull request by tomlm**
Fix for #196 
accessText

### CODE REVIEW

File: src/Consolonia.Core/Controls/ConsoloniaAccessText.cs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+            switch (e.Property.Name)
+            {
+                case nameof(Text):
+                    {
+                        if (!String.IsNullOrEmpty(Text))
+                        {
+                            _accessRun = null;
+                            InlineCollection inlines = new InlineCollection();
+                            var iPos = Text.IndexOf('_', StringComparison.Ordinal);
+                            if (iPos >= 0 && iPos < Text.Length - 1)
```
Comment (jinek) [inline src/Consolonia.Core/Controls/ConsoloniaAccessText.cs:32]: I think complex condition can be replaced with just iPos!=-1

File: src/Consolonia.Core/Controls/ConsoloniaAccessText.cs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+                case nameof(Text):
+                    {
+                        if (!String.IsNullOrEmpty(Text))
+                        {
+                            _accessRun = null;
+                            InlineCollection inlines = new InlineCollection();
+                            var iPos = Text.IndexOf('_', StringComparison.Ordinal);
+                            if (iPos >= 0 && iPos < Text.Length - 1)  
+                            {
+                                inlines.Add(new Run(Text.Substring(0, iPos)));
```
Comment (jinek) [inline src/Consolonia.Core/Controls/ConsoloniaAccessText.cs:34]: Just a hint: Text[..iPos] or Text[iPos..]

File: src/Consolonia.Core/Controls/ConsoloniaAccessText.cs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+        private void OnPropertyChanged(object sender, Avalonia.AvaloniaPropertyChangedEventArgs e)
+        {
+            switch (e.Property.Name)
+            {
+                case nameof(Text):
+                    {
+                        if (!String.IsNullOrEmpty(Text))
+                        {
+                            _accessRun = null;
```
Comment (jinek) [inline src/Consolonia.Core/Controls/ConsoloniaAccessText.cs:29]: Looks not necessary

**Review by jinek — COMMENTED** [review]
Button and other controls use some kind of factory or template, which can be also overridden to use ConsoloniaAccessText.

File: src/Consolonia.Core/Controls/ConsoloniaAccessText.cs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+                                _accessRun.TextDecorations.Add(new TextDecoration { Location = TextDecorationLocation.Underline });
+                            else
+                                _accessRun.TextDecorations.Clear();
+                        }
+                        break;
+                    default:
+                        break;
+                }
+
+            };
```
[the remark this answers is not in the document]
Comment (jinek) [reply src/Consolonia.Core/Controls/ConsoloniaAccessText.cs]: How can the memory leak be caused if the event is instance, not static?

[the remark this answers is not in the document]
Comment (jinek) [reply src/Consolonia.Core/Controls/ConsoloniaAccessText.cs]: @tomlm do we use it? Thinking again to switch it off. It sometimes seems we are heating earth for nothing.

**Comment by tomlm** [comment]
I was looking, couldn't find it. Any idea where?

File: src/Consolonia.Core/Controls/ConsoloniaAccessText.cs (hunk 2 above)
Comment (tomlm) [reply src/Consolonia.Core/Controls/ConsoloniaAccessText.cs:34]: oh, fancy. I totally forgot about that new feature. It hasn't been integrated into my coding patterns.

Comment (tomlm) [reply src/Consolonia.Core/Controls/ConsoloniaAccessText.cs:34]: Holy shit, I just tried it...so clean, so readable, so succinct. .Substring always felt so verbose. I love it!  Thanks for the tip!

**Comment by tomlm** [comment]
it works almost.
There are 2 issues left:
**Accelerator used for Label puts focus on the input control AND puts accelerator key into the input**
label
Alt+F puts focus on the target element, but then even though I set the e.Handled=true it puts the letter 'F' into the input control. 
It should just set the focus and not add the 'F'.

**Toggle of Alt accelerator works only after the FIRST time I use Menu**
This is super weird. 
On Avalonia if you hit alt key, it shows the underscore, then hit Alt key again it hides it. If you hit the accelerator, it activates the target and hides it.
In Consolonia, if you hit the Alt key it shows the underscore, then hit Alt key again it leaves it shown. 
Unless I go to the Menu page. When I go to the menu page, then it starts working, and it works until I shut down the app. 
WTF???

label2

**Comment by tomlm** [comment]
The second is an avalonia bug. I just repoed in avalonia proper.

**Comment by jinek** [comment]
First one I can check.
Do I need to review now? Lets please use draft PRs to indicate it's not for review in case it's not. It's sometimes hard for me to prioritize what to do here.

File: src/Consolonia.Core/Controls/ConsoloniaAccessText.cs [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+                case nameof(Text):
+                {
+                    if (!string.IsNullOrEmpty(Text))
+                    {
+                        var inlines = new InlineCollection();
+                        int iPos = Text.IndexOf('_', StringComparison.Ordinal);
+                        if (iPos >= 0 && iPos < Text.Length - 1)
+                        {
+                            inlines.Add(new Run(Text[..iPos]));
+                            _accessRun = new Run(Text[++iPos..++iPos]);
```
Comment (jinek) [inline src/Consolonia.Core/Controls/ConsoloniaAccessText.cs:38]: it looks stunning!)

**Comment by jinek** [comment]
> it works almost. There are 2 issues left: **Accelerator used for Label puts focus on the input control AND puts accelerator key into the input**

Could you please push the repro of label+input somewhere?

**Comment by tomlm** [comment]
It should be pushed....there should be a panel page in gallery

**Comment by tomlm** [comment]
I'm ready to check it in. Sorry. I was ready, then based on feedback on avalonia side I made some changes, and then feedback from you I added the Label/Button support. etc. 

I will try to be more disclipined about using Draft PR. Sorry.

File: src/Consolonia.Core/Controls/ConsoloniaAccessText.cs (hunk 5 above)
Comment (tomlm) [reply src/Consolonia.Core/Controls/ConsoloniaAccessText.cs:38]: 🤪

**Comment by jinek** [comment]
> I'm ready to check it in. Sorry. I was ready, then based on feedback on avalonia side I made some changes, and then feedback from you I added the Label/Button support. etc.
> 
> I will try to be more disclipined about using Draft PR. Sorry.

no worries. But I want to investigate ALT handling also, the guy you've changed.

Thought I have investigated the issue of producing excessive F.
Issue happens because `ConsoleWindow.ConsoleOnKeyEvent` produces `RawTextInputEventArgs` after the focus was already set because previously `RawKeyEventArgs.Handled` was not set to true by AccessKey infrastructure.
It looks like either `Dispatcher.UIThread.InvokeAsync` does not wait for the delegate to complete or may be this logic has been changed completely. I will check if Avalonia has reworked this part, we may be just need to reflect those changes from new version of Avalonia.

**Comment by tomlm** [comment]
I bet we should be awaiting the InvokeAsync. There are 2 patterns, there is Post() and Invoke(). Invoke returns a task, we should be careful about the semantics of which one we are using.

**Comment by jinek** [comment]
> I bet we should be awaiting the InvokeAsync. There are 2 patterns, there is Post() and Invoke(). Invoke returns a task, we should be careful about the semantics of which one we are using.

This one excluded. Task returned from Invoke properly awaits for access key to be processed.

I could not find matching code in Avalonia at all. Thinking may be I just invented this processing: if KeyDown event was not handled, then producing TextInput event.
I will try to do the trick by checking if a key was pressed together with ALT.

**Comment by jinek** [comment]
Yeah, that helped. fixed first issue

**Comment by tomlm** [comment]
Second issue is Avalonia bug. It looks like we can work around by having an empty MENU on the page.

**Comment by jinek** [comment]
@tomlm 
please, check my last commit. I have removed special handling of ALT key. It seems all works fine without

**Comment by tomlm** [comment]
Looks good.

**Review by jinek — APPROVED** [review]

**Comment by tomlm** [comment]
Thanks Evgeny! 

If you are ever in Seattle I'll buy you a beer!

**Comment by jinek** [comment]
No, I will buy you a beer. 😃

**Pull request merged**